src/Stratis.Bitcoin.Features.Consensus/CoinViews/DBreezeCoinView.cs

```diff
@@ -319,49 +322,92 @@ public Task<RewindData> GetRewindData(int height)
-
-                        foreach (uint256 txId in rewindData.TransactionsToRemove)
+                        for (int height = firstBlockToRewind; height <= currentHeight; height++)
                         {
-                            this.logger.LogTrace("Outputs of transaction ID '{0}' will be removed.", txId);
-                            transaction.RemoveKey("Coins", txId.ToBytes(false));
+                            Row<int, byte[]> row = transaction.Select<int, byte[]>("Rewind", height);
+                            if (!row.Exists)
+                                continue;
+
+                            keysToRemove.Add(row.Key);
+                            var rewindData = this.dBreezeSerializer.Deserialize<RewindData>(row.Value);
+
+                            if (height == (targetHeight + 1))
+                                firstRewindData = rewindData;
+
+                            foreach (uint256 txId in rewindData.TransactionsToRemove)
+                            {
+                                this.logger.LogTrace("Outputs of transaction ID '{0}' will be removed.", txId);
+                                transactionsToRemove.Add(txId);
+                            }
+
+                            foreach (UnspentOutputs coin in rewindData.OutputsToRestore)
+                            {
+                                this.logger.LogTrace("Outputs of transaction ID '{0}' will be restored.", coin.TransactionId);
+                                outputsToRestore.Add(coin);
+                            }
                         }
 
-                        foreach (UnspentOutputs coin in rewindData.OutputsToRestore)
+                        if (keysToRemove.Count > 0)
                         {
-                            this.logger.LogTrace("Outputs of transaction ID '{0}' will be restored.", coin.TransactionId);
-                            transaction.Insert("Coins", coin.TransactionId.ToBytes(false), this.dBreezeSerializer.Serialize(coin.ToCoins()));
+                            transaction.SynchronizeTables("BlockHash", "Coins", "Rewind");
+
+                            var byteListComparer = new ByteListComparer();
+
+                            foreach (int key in keysToRemove.OrderBy(k => k))
+                                transaction.RemoveKey("Rewind", key);
+
+                            foreach (uint256 txId in transactionsToRemove.OrderBy(t => t.ToBytes(false), byteListComparer))
+                                transaction.RemoveKey("Coins", txId.ToBytes(false));
+
+                            foreach (UnspentOutputs coin in outputsToRestore.OrderBy(c => c.TransactionId.ToBytes(false), byteListComparer))
+                                transaction.Insert("Coins", coin.TransactionId.ToBytes(false), this.dBreezeSerializer.Serialize(coin.ToCoins()));
+
+                            this.SetBlockHash(transaction, firstRewindData.PreviousBlockHash);
+
+                            transaction.Commit();
```

Comment:
here the Commit is called inside the loop, it seems to me that for every height we have to rollback, we create a transaction and then commit it.
Why not having just one single commit and one single transaction? basically why not create a transaction outside the `while (targetHeight < currentHeight)` and commit the transaction at the end?

Reply:
and another improvement that can be done, is to synchronize only tables that are affected by the operation
`transaction.SynchronizeTables("BlockHash", "Coins", "Rewind");`

there are cases where we have just to update BlockHash table on startup because the block tip is higher than consensus and we have nothing to restore.
In this case we can get rid of SynchronizeTables because for changes that applies on a single table this can be avoided (=performance improvement)

Reply:
remind: It's important to remember that `transaction.SynchronizeTables` must be called before any table modification (it's fine to have selects before)

Reply:
1) The purpose of having a batch limit is to allow control over the size that these collections can grow:
```
var transactionsToRemove = new List<uint256>();
var outputsToRestore = new List<UnspentOutputs>();
```
2) Keeping the code simple instead of a making a very small performance improvement.
3) The current code is doing that for each batch.

Reply:
It's fine I didn't saw MaxBlocksToRewind at first

Reply:
Fortunately we are only recording the transaction hashes.